contracts/governance/Vesting/VestingLogic.sol

```diff
@@ -110,68 +120,75 @@ contract VestingLogic is IVesting, VestingStorage, ApprovalReceiver {
      * */
-    function governanceWithdrawTokens(address receiver) public {
-        require(msg.sender == address(staking), "unauthorized");
-
-        _withdrawTokens(receiver, true);
+    function withdrawTokens(address _receiver) public onlyOwners {
+        _withdrawTokens(_receiver, 0);
     }
 
     /**
-     * @notice Withdraws unlocked tokens from the staking contract and
+     * @notice Withdraws unlocked tokens from the staking contract from the given start time (for optimization) and
      * forwards them to an address specified by the token owner.
-     * @param receiver The receiving address.
+     *
+     * @param _receiver The receiving address.
+     * @param  _startFrom The start value for the iterations.
      * */
-    function withdrawTokens(address receiver) public onlyOwners {
-        _withdrawTokens(receiver, false);
+    function withdrawTokensStartingFrom(address _receiver, uint256 _startFrom) public onlyOwners {
+        _withdrawTokens(_receiver, _startFrom);
     }
 
     /**
      * @notice Withdraws tokens from the staking contract and forwards them
      * to an address specified by the token owner. Low level function.
      * @dev Once here the caller permission is taken for granted.
-     * @param receiver The receiving address.
-     * @param isGovernance Whether all tokens (true)
+     * @param _receiver The receiving address.
+     * @param _startFrom The start value for the iterations.
      * or just unlocked tokens (false).
      * */
-    function _withdrawTokens(address receiver, bool isGovernance) internal {
-        require(receiver != address(0), "receiver address invalid");
+    function _withdrawTokens(address _receiver, uint256 _startFrom) internal {
+        require(_receiver != address(0), "receiver address invalid");
 
         uint96 stake;
 
         /// @dev Usually we just need to iterate over the possible dates until now.
         uint256 end;
 
+        uint256 defaultStart = startDate + cliff;
+
+        _startFrom = _startFrom >= defaultStart ? _startFrom : defaultStart;
+
         /// @dev In the unlikely case that all tokens have been unlocked early,
         ///   allow to withdraw all of them.
-        if (staking.allUnlocked() || isGovernance) {
+        if (staking.allUnlocked()) {
             end = endDate;
         } else {
             end = block.timestamp;
         }
 
+        uint256 totalIterationValue =
+            (_startFrom + (FOUR_WEEKS * getMaxVestingWithdrawIterations()));
+        uint256 adjustedEnd = end < totalIterationValue ? end : totalIterationValue;
+
         /// @dev Withdraw for each unlocked position.
         /// @dev Don't change FOUR_WEEKS to TWO_WEEKS, a lot of vestings already deployed with FOUR_WEEKS
         ///		workaround found, but it doesn't work with TWO_WEEKS
-        for (uint256 i = startDate + cliff; i <= end; i += FOUR_WEEKS) {
+        for (uint256 i = _startFrom; i < adjustedEnd; i += FOUR_WEEKS) {
```

Comment:
```suggestion
    for (uint256 i = _startFrom; i <= adjustedEnd; i +=   FOUR_WEEKS) {
```
also pls add test to watch for this corner case

Reply:
Shoud be `<`
Using `<=` will cause the 1 surplus iteration.

Reply:
it used to be `<= end`; now the `end` is just adjusted considering max iterations 
whay is it now `<`?

Reply:
As i mentioned above, using <= will have additional iteration.
Illustration: 
startFrom: 8
increment = 4
MaxIterations: 5

Adjusted end would be (8 + (4*5)) = 28

Iteration 1 = 8
Iteration 2 = 12
Iteration 3 = 16
Iteration 4 = 20
Iteration 5 = 24

If we use `<=` , then it will have 6 iterations instead of 5.

Reply:
it is only true in the case of adjustedEnd == totalIterationValue  
it will not work if adjustedValue == end  
e.g.  
adjustedEnd == end == (startFrom + 2*4wks periods) < maxIterations which will be fair in the most cases  
then iterations:  
1) i = startFrom -> OK
2) i = startFrom + 4 weeks -> OK
3) i = startFrom + 8 weeks will be lost 'cause i < adjustedEnd
=============================================================
i recommend to replace 
```
uint256 totalIterationValue = (_startFrom + (FOUR_WEEKS * getMaxVestingWithdrawIterations()));
```  
with
```
uint256 totalIterationValue = (_startFrom + (FOUR_WEEKS * (getMaxVestingWithdrawIterations() - 1)));
```  
as the first iteration is actually 0 indexed, i.e. _startFrom
which will not cause overflow as you have require for `getMaxVestingWithdrawIterations()` to be > 1 when setting the value.  
also:
- pls adjust tests because they should not have passed for the described case.  
- setting the max iterations should a part of the deployment process (take a note on this when you'll be adding it to the hh Staking modules deployment script)

Reply:
Good point
fixed in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/447/commits/3ad39d89859e2587e4371943783db45d81164580)